co-log/src/Colog/Rotation.hs

```diff
@@ -0,0 +1,112 @@
+module Colog.Rotation
+       (
+         Limit(..)
+       , withLogRotation
+       ) where
+
+import Data.Maybe (mapMaybe)
+import Data.Semigroup (Max (..))
+import System.IO (hFileSize)
+
+import Colog.Core.Action (LogAction (..))
+
+import qualified System.Directory as D
+import qualified System.FilePath.Posix as POS
+
+
+data Limit = LimitTo Natural | Unlimited deriving (Eq, Ord)
+
+{- | Logger rotation action. Takes name of the logging file @file.foo@. Always
+writes new logs to file named @file.foo@ (given file name, also called as /hot log/).
+
+* If the size of the file exceeds given limit for file sizes then this action
+  renames @file.foo@ to @file.foo.(n + 1)@ (where @n@ is the number of latest
+  renamed file).
+* If the number of files on the filesystem is bigger than the files number limit
+  then the given @FilePath -> IO ()@ action is called on the oldest file. As
+  simple solution, you can pass @removeFile@ function to delete old files but
+  you can also pass some archiving function if you don't want to lose old logs.
+-}
+withLogRotation
+    :: forall r msg m .
+       MonadIO m
+    => Limit
+    -- ^ Max allowed file size in bytes
+    -> Limit
+    -- ^ Max allowed number of files to have
+    -> FilePath
+    -- ^ File path to log
+    -> (FilePath -> IO ())
+    -- ^ What to do with old files; pass @removeFile@ here for deletion
+    -> (Handle -> LogAction m msg)
+    -- ^ Action that writes to file handle
+    -> (LogAction m msg -> IO r)
+    -- ^ Continuation action
+    -> IO r
+withLogRotation sizeLimit filesLimit path cleanup mkAction cont = do
+    -- TODO: figure out how to use bracket to safely manage
+    -- possible exceptions
+    handle <- openFile path AppendMode
+    handleRef <- newIORef handle
+    cont $ rotationAction handleRef
+  where
+    rotationAction :: IORef Handle -> LogAction m msg
+    rotationAction refHandle
+      = LogAction $ \msg -> do
+        handle <- liftIO $ readIORef refHandle
+        unLogAction (mkAction handle) msg
+
+        whenM
+          (liftIO $ isFileSizeLimitReached sizeLimit handle)
```

Comment:
You can put `liftIO` in front of `do` in `isFileSizeLimitReached` function to make this `whenM` call look slightly nicer.